### DISCUSSION TITLE
Select the right service for dependency injection

### DISCUSSION THREAD

**Discussion by CJDennis** [symfony/symfony#42494] [Q&A] [answered]
I'm converting some 3.4 code to run under 4.0. I have this line in a Controller:
```
$transport = $this->get('swiftmailer.transport.real');
```

In 3.4 it fetches a `Swift_Transport_EsmtpTransport` object, but it throws an error in 4.0:
```
The "swiftmailer.transport.real" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.
```

If I try to use dependency injection in 4.0 with `Swift_Transport`, it injects a `Swift_Transport_SpoolTransport` object instead of a `Swift_Transport_EsmtpTransport` object which then exhausts all available memory on the very next line.
```
$spool->flushQueue($transport);
```

```
Fatal error: Allowed memory size of 1073741824 bytes exhausted (tried to allocate 20480 bytes)
```

How do I use dependency injection to get the right object?

These are the relevant lines:
```
class ... extends Controller {
    /** @var Swift_Mailer */
    protected $swift_mailer;
    /** @var Swift_Transport */
    protected $swift_transport;

    public function __construct(Swift_Mailer $swift_mailer, Swift_Transport $swift_transport) {
        $this->swift_mailer = $swift_mailer;
        $this->swift_transport = $swift_transport;
    }

    function ... {
            //$transport = $this->get('mailer')->getTransport(); // Works in 3.4
            $transport = $this->swift_mailer->getTransport(); // Works in 4.0
            $spool = $transport->getSpool();

            $transport = $this->get('swiftmailer.transport.real'); // Works in 3.4
            //$transport = $this->swift_transport; // Doesn't work in 4.0
            $spool->flushQueue($transport);
```

**Comment by romaricdrigon** [comment, marked as answer]
I don't know if there may be something SwiftMailer specific, but you could use a `bind`: it allows you to bind a specific service instance to a parameter name. Something quite useful when upgrading from legacy!

In your `services.yaml`, do:
```yaml
services:
    _defaults:
        bind:
            $swiftMailerTransport: '@swiftmailer.transport.real'
```

Now in your controller you can:
```php
public function (Swift_Transport_EsmtpTransport $swiftMailerTransport)
{
    // ...
```

**Reply by CJDennis** [reply]
Thanks! But I got this error:
```
[TypeError] Argument 2 passed to HMRX\AppBundle\Controller\PharmacistController::__construct() must be an instance of Swift_Transport_EsmtpTransport, instance of Swift_Transport_NullTransport given
```

**Reply by romaricdrigon** [reply]
Right. Then I suppose SwiftMailer bundle is doing some work at compilation, then this service is not correctly set.
Then let's try another approach: instead of a bind, configuring DI to inject the correct service to your controller.

```yaml
services:
    App\Controller\YourController:
        arguments:
            $swiftMailerTransport: '@swiftmailer.transport.real'
```
Though likely you should keep `Swift_Transport` parameter type hint, as at some point another transport could be passed.

**Reply by CJDennis** [reply]
That still passes in a `Swift_Transport_NullTransport` object. Has `@swiftmailer.transport.real` been changed to a different service in 4.0, or perhaps it's different under test?

**Reply by romaricdrigon** [reply]
Oh right, you are in "test" environment? Then please check your configuration, you likely have `disabled_delivery` set to `false`. Hence the transport is replaced by a no-op.

**Reply by CJDennis** [reply]
Well, I don't actually want to send real emails in test, I'm writing a test for an old piece of code that hasn't been covered yet. Obviously it's really important that it works in both test and production! As I mentioned in my original post, the test runs out of memory when I try to use dependency injection. https://symfony.com/doc/current/email.html looks useful for testing that the email has the correct contents, but it hasn't solved my problem.

**Reply by CJDennis** [reply]
OK, I think that's actually done the trick! I forgot I'd put in a `throw` to see what class was being passed in! Now that I've taken it out, it seems to be working! That is, the memory issue is solved, and I've got a completely unrelated fatal bug elsewhere in my code now!
